File: Android/bet/src/main/java/com/felipearpa/tyche/bet/pending/PendingBetItemViewState.kt
package com.felipearpa.tyche.bet.pending

sealed class PendingBetItemViewState(open val value: PartialPoolGamblerBetModel) {
    data class Visualization(override val value: PartialPoolGamblerBetModel) :
        PendingBetItemViewState(value)

    data class Edition(override val value: PartialPoolGamblerBetModel) :
        PendingBetItemViewState(value)

    companion object
}

fun PendingBetItemViewState.Companion.emptyVisualization(): PendingBetItemViewState {
    return PendingBetItemViewState.Visualization(emptyPartialPoolGamblerBetModel())
}

fun PendingBetItemViewState.copy(value: PartialPoolGamblerBetModel): PendingBetItemViewState {
    return when (this) {
        is PendingBetItemViewState.Visualization -> PendingBetItemViewState.Visualization(
            value
        )

        is PendingBetItemViewState.Edition -> PendingBetItemViewState.Edition(value)
    }
}

File: iOS/Bet/Sources/Bet/Pending/PendingBetItemViewState.swift
enum PendingBetItemViewState {
    case visualization(_ value: PartialPoolGamblerBetModel)
    case edition(_ value: PartialPoolGamblerBetModel)
    
    var value: PartialPoolGamblerBetModel {
        get {
            switch self {
            case .visualization(let value): value
            case .edition(let value): value
            }
        }
        set {
            switch self {
            case .visualization: self = .visualization(newValue)
            case .edition: self = .edition(newValue)
            }
        }
    }
}

extension PendingBetItemViewState {
    func copy(value: PartialPoolGamblerBetModel) -> PendingBetItemViewState {
        switch self {
        case .visualization: .visualization(value)
        case .edition: .edition(value)
        }
    }
}

extension PendingBetItemViewState {
    static func emptyVisualization() -> PendingBetItemViewState {
        .visualization(PartialPoolGamblerBetModel.empty())
    }
}
